test/Generics/requirement_machine_diagnostics.swift

```diff
@@ -197,4 +197,109 @@ func inferred6<T : P11>(_: T) where T.Y : Hashable, T.Z == Set<T.X>, T.X == T.Y
 
 func typeMatcherSugar<T>(_: T) where Array<Int> == Array<T>, Array<Int> == Array<T> {}
 // expected-warning@-1 2{{redundant same-type constraint 'Array<Int>' == 'Array<T>'}}
-// expected-warning@-2{{redundant same-type constraint 'T' == 'Int'}}
\ No newline at end of file
+// expected-warning@-2{{redundant same-type constraint 'T' == 'Int'}}
+
+// MARK: - Conflict diagnostics
+
+protocol ProtoAlias1 {
+  typealias A1 = Int
+}
+
+protocol ProtoAlias2 {
+  typealias A2 = String
+}
+
+func basicConflict<T: ProtoAlias1 & ProtoAlias2>(_:T) where T.A1 == T.A2 {}
+// expected-error@-1{{generic signature requires types 'Int' and 'String' to be the same}}
+
+protocol RequiresAnyObject {
+  associatedtype A: AnyObject
+}
+
+protocol RequiresConformance {
+  associatedtype A: P
+}
+
+class Super {}
+protocol RequiresSuperclass {
+  associatedtype A: Super
+}
+
+func testMissingRequirements() {
+  struct S {}
+  func conflict1<T: RequiresAnyObject>(_: T) where T.A == S {}
+  // expected-error@-1{{same-type constraint type 'S' does not conform to required protocol 'AnyObject'}}
+
+  func conflict2<T: RequiresConformance>(_: T) where T.A == C {}
+  // expected-error@-1{{same-type constraint type 'C' does not conform to required protocol 'P'}}
+
+  class C {}
+  func conflict3<T: RequiresSuperclass>(_: T) where T.A == C {}
+  // expected-error@-1{{type 'T.A' cannot be a subclass of both 'Super' and 'C'}}
+
+  func conflict4<T: RequiresSuperclass>(_: T) where T.A: C {}
+  // expected-error@-1{{type 'T.A' cannot be a subclass of both 'Super' and 'C'}}
+}
+
+protocol Fooable {
+  associatedtype Foo
+
+  var foo: Foo { get }
+}
+
+protocol Barrable {
+  associatedtype Bar: Fooable
+  var bar: Bar { get }
+}
+
+func sameTypeConflicts() {
```

Comment:
I like using local types/functions nested inside of another function to group tests!

Can you add a couple of tests for conflicts inside protocol `where` clauses as well just in case there's something wrong with the getTypeForTerm() usage?